parser-util/main.cc

```diff
@@ -0,0 +1,346 @@
+/* ========================================================================== *
+ *
+ * Parse various URIs and junk using `nix' libraries and expose them in a
+ * trivially simple way so that they can be used consumed by other software.
+ *
+ * -------------------------------------------------------------------------- */
+
+#include <stddef.h>
+#include <iostream>
+#include <string>
+#include <cstring>
+#include <nlohmann/json.hpp>
+#include <nix/shared.hh>
+#include <nix/eval.hh>
+#include <nix/eval-inline.hh>
+#include <nix/flake/flake.hh>
+#include <nix/store-api.hh>
+#include <cassert>
+#include <ranges>
+
+
+/* -------------------------------------------------------------------------- */
+
+static nix::flake::LockFlags floxFlakeLockFlags = {
+  .updateLockFile = false
+, .writeLockFile  = false
+, .applyNixConfig = false
+};
+
+
+/* -------------------------------------------------------------------------- */
+
+namespace nix {
+    void
+  to_json( nlohmann::json & j, const FlakeRef & ref )
+  {
+    j = { { "string", ref.to_string() }
+        , { "attrs",  fetchers::attrsToJSON( ref.toAttrs() ) }
+        };
+  }
+}  /* End namespace `nix' */
+
+
+/* -------------------------------------------------------------------------- */
+
+  static nlohmann::json
+parseURI( const char * arg )
+{
+  try
+    {
+      nix::ParsedURL       url    = nix::parseURL( arg );
+      nix::ParsedUrlScheme scheme = nix::parseUrlScheme( url.scheme );
+
+      nlohmann::json schemeJSON = {
+        { "full",        url.scheme       }
+      , { "application", nlohmann::json() }
+      , { "transport",   scheme.transport }
+      };
+      if ( scheme.application.has_value() )
+        {
+          schemeJSON["application"] = scheme.application.value();
+        }
+
+      nlohmann::json uriJSON = {
+        { "base",      std::move( url.base )     }
+      , { "scheme",    std::move( schemeJSON )   }
+      , { "authority", nlohmann::json()          }
+      , { "path",      std::move( url.path )     }
+      , { "fragment",  std::move( url.fragment ) }
+      , { "query",     std::move( url.query )    }
+      };
+      if ( url.authority.has_value() )
+        {
+          uriJSON["authority"] = url.authority.value();
+        }
+
+      return uriJSON;
+    }
+  catch( std::exception & e )
+    {
+      std::cerr << e.what() << std::endl;
+      exit( EXIT_FAILURE );
+    }
+
+  /* Unreachable */
+  assert( false );
+  return nlohmann::json();
```

Comment:
Should we still have code here if you believe it's unreachable?

Reply:
Yeah its to keep the compiler happy. It requires that something be returned in all conditional branches, so even though I believe that block is unreachable I still have to return something.

The assert helps avoid runtime misbehaving if my belief about the block being unreachable was incorrect.

Reply:
If we're using C++23 we can even do something fancy. 😉
https://en.cppreference.com/w/cpp/utility/unreachable